### PR TITLE
Add Kafka tutorial.

### PR DESCRIPTION
**Pull request by htroisi**
Signed-off-by: Hannah Troisi <htroisi@pixielabs.ai>

### CODE REVIEW

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+metaTitle: "Tutorials | Pixie 101 | Kafka Monitoring"
+metaDescription: "Learn how to monitor Kafka using Pixie."
+order: 7
+---
+
+Debugging distributed messaging systems can be challenging. Pixie makes analyzing Kafka easier by using eBPF to [automatically capture](https://docs.px.dev/about-pixie/pixie-ebpf/) full-body Kafka requests without the need for manual instrumentation.
+
+This tutorial will demonstrate how to use Pixie to see:
+
+- [Flow graph of topic producers and consumers](#topic-flow-graph)
```
Comment (chengruizhe) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: Topic-centric flow graph and topic summaries

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+- [Latency and throughput per pod](#latency-and-throughput-per-pod)
+- [Full-body Kafka requests](#full-body-kafka-requests)
```
Comment (chengruizhe) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: Kafka events with parsed body or Kafka events with metadata?
Don't want to say full-body tracing, since we don't actually show the payload of the events. We do show everything important in the control-plane of Kafka.

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+- [Producer-consumer latency per topic](#producer-consumer-latency)
```
Comment (chengruizhe) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: Producer-consumer latency per topic and partition

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+> - Navigate to the external IP (with port) in your browser.
+> - Press the `"Start swarming"` button.
+
+## Kafka Microservice Demo App
+
+The demo application you deployed in the [Prerequisites](#prerequisites) steps has:
+
+- A single `order` topic.
+- One producer: the `order` service publishes messages to the `order` topic.
+- Two consumers: the `shipping` and `invoicing` services consume messages from the `order` topic. The `shipping` service is able consume the order messages at the rate they are produced. The `invoicing` service is slower and unable to consume the order messages at the rate they are produced.
```
Comment (chengruizhe) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: The `shipping` service is able **to** consume

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+<svg title='' src='use-case-tutorials/kafka/kafka-overview.png'/>
+:::
+
+> This graph shows us that we have 1 topic, named `order`, with 1 producer and 2 consumers.
+
+<Alert variant="outlined" severity="info">If this script doesn't show any output, make sure that you started the load generator in Prerequisites Step #3.</Alert>
+
+2. Hover over an edge on the graph to see throughput and total record Bytes for a producer or consumer. The thickness of the edge indicates an increase in throughput.
+
+> By examining the edges of the graph, we can see that the producer is producing to the order topic at about 450 B/s. The `shipping` consumer is consuming at the same rate of 450 B/s. The `invoicing` consumer is only consuming at 86 B/s and is falling behind.
```
Comment (chengruizhe) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: 450 B/s is a really low number for Kafka applications. We may want to say something like: Since this is a demo application, the traffic throughput shown here is lower than normal.

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+> The table at the bottom summarizes the same topic information, but also includes the number of topic partitions.
+
+## Latency and Throughput per Pod
+
+The previous script alerted us to the fact that the `invoicing` pod is consuming messages at a slower speed than they are being produced. Let's investigate this more.
```
Comment (chengruizhe) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: at a **lower** speed or more slowly

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+1. Select `px/kafka_stats` from the script drop-down menu.
+
+> This script calculates the latency and throughput of a pod's Kafka requests.
+
+::: div image-xl relative
+<svg title='' src='use-case-tutorials/kafka/kafka-stats.png'/>
+:::
+
+> The **Summary** table at the bottom of the page shows us the request throughput, latency, and total number of Kafka requests per pod. The `shipping` pod has sent around 7,600 `Fetch` requests while the `invoicing` pod has sent much fewer (only 1700 `Fetch` requests). We can also see that the This again indicates that something is wrong.
```
Comment (chengruizhe) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: `We can also see that ...`. Need to complete this sentence.

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+2. To check the pod's resources, click on the `px-kafka/invoicing-*` pod name in the **SOURCE** column of the **Summary** table to follow the deep link to the `px/pod` script.
+
+> This script shows an overview of the pod's CPU usage, network traffic and throughput, disk and memory usage.
+
+> In this case, CPU utilization is low, which means resource starvation is not causing the `invoicing` pod to consume messages at a slower rate.
+
+## Full-body Kafka Requests
```
Comment (chengruizhe) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: Kafka events with parsed body

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+Let's inspect the raw Kafka requests flowing through the cluster.
+
+1. Select `px/kafka_data` from the script drop-down menu.
+
+> This script shows a sample of the most recent Kafka messages in your cluster.
+
+> For each record you can see the source, destination, request command, full-body request and response, and the latency.
```
Comment (chengruizhe) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: Maybe a note saying that: We only show size of the payload, not the content, because it's usually too large.

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+
+> The graph updates to show non-zero latency for all 5 `invoicing` partitions. All 5 partitions are consuming `order` topic messages around 40-55 seconds after they are being produced.
+
+> The zig zag line shape is due to the `invoicing` consumer fetching a batch of messages from the topic, processing these messages slowly, then fetching another group of messages and being even more behind. The overall trend is upward; there is increasing producer-consumer lag for the `invoicing` consumer. This is a bad sign.
+
+## Consumer Rebalancing Delay
+
+Consumer rebalancing (shuffling the consumers among partitions) happens whenever a new consumer comes online or an existing consumer goes offline.
+
+It's important to monitor these events because either some or all of the consumers are stopped from consuming messages when the rebalancing is in progress. If consumer rebalancing events happen too often, this can cause consumers to lag in their consumption of topics.
```
Comment (chengruizhe) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: consumption of messages.

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+
+> This script visualizes the most recent Kafka consumer rebalancing events with their delays.
+
+::: div image-xl relative
+<svg title='' src='use-case-tutorials/kafka/kafka-consumer-rebalancing.png'/>
+:::
+
+> The **Consumer Groups** table shows us that we now have 3 members in the `shipping` consumer group.
+
+> Each consumer rebalancing event consists of 1 `JoinGroup` and 1 `SyncGroup` request per consumer. The **Consumer Rebalancing Events** table shows the traced `JoinGroup` and `SyncGroup` requests.
```
Comment (chengruizhe) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: This may not strictly be true, depending on the consumer rebalancing strategy in use. 
May want to add: Each consumer rebalancing event consists of 1 `JoinGroup` and 1 `SyncGroup` request per consumer. In the default rebalance protocol, there's a stop-the-world effect on the whole consumer group when a rebalancing event happens.

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+
+> The **Consumer Rebalancing Delay** table at the top calculates the delay between when the `JoinGroup` request was sent and when the `SyncGroup` request was received.
```
Comment (chengruizhe) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: This measures the stop-the-world delay mentioned above.

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+> - Run `kubectl get pods -n px-kafka` to make sure all pods are ready before proceeding.
+> - Run `px demo interact px-kafka` and follow the directions to turn on the invoicing service delay.
+
+## Kafka Microservice Demo App
+
+The demo application you deployed in the [Prerequisites](#prerequisites) section has:
+
+- A single `order` topic.
+- One producer: the `order` service publishes messages to the `order` topic.
+- Two consumers: the `shipping` and `invoicing` services consume messages from the `order` topic. The `shipping` service is able to consume the order messages at the rate they are produced. The `invoicing` service is slower and unable to consume the order messages at the rate they are produced.
```
Comment (oazizi000) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: Maybe include a comment tying this back to the `px demo interact px-kafka step`...could be as simple as just changing the phrasing to:

The invoicing service, when configured with the delay above, is slower...

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+<svg title='' src='use-case-tutorials/kafka/kafka-microservice-app.png'/>
+:::
+
+To see the application's simple front end, navigate to the external IP (and port) of the `apache` service:
+
+```bash
+kubectl -n px-kafka get svc apache
+```
+
+To use the application, click on the Order page and scroll to the bottom of the page to select the `"Add Order"` button. Any orders added on the Order page should soon show up on both the Shipping and Invoicing pages.
```
Comment (oazizi000) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: If the load gen is running, this may be hard to follow. Can we add a comment that the load gen is also generating orders so the reader is not confused.

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+
+2. Select the drop-down arrow next to the `namespace` argument, type `px-kafka`, and press `Enter` to re-run the script.
+
+::: div image-xl relative
+<svg title='' src='use-case-tutorials/kafka/kafka-overview.png'/>
+:::
+
+> Pixie is able to automatically discover and list all clients and topics in the cluster.
+
+> This graph shows us that our microservice app has 1 topic, named `order`, with 1 producer and 2 consumers.
```
Comment (oazizi000) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: Can we clarify that this only covers _active_ topics?

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+
+4. Scroll down the page to see tables listing the brokers, producer and consumer pods.
+
+::: div image-xl relative
+<svg title='' src='use-case-tutorials/kafka/kafka-overview-2.png'/>
+:::
+
+> The **Kafka Producer Pods** and **Kafka Consumer Pods** tables confirm that the `invoicing` client is consuming messages at a slower speed than they are being produced.
+
+> The `shipping` client has sent 292 `Fetch` requests while the `invoicing` client has sent much fewer (only 192 `Fetch` requests). This again indicates that something is wrong.
```
Comment (oazizi000) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: "In our case, the `shipping` client has..."

to clarify that their numbers might be different.

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+
+> In this case, CPU utilization is low, which means resource starvation is not causing the `invoicing` client to consume messages at a slower rate.
+
+## Kafka Events with Metadata
+
+Let's inspect the raw Kafka requests flowing through the cluster.
+
+1. Select `px/kafka_data` from the script drop-down menu.
+
+> Pixie is able to [automatically trace](https://docs.px.dev/about-pixie/pixie-ebpf/) all messages flowing through your cluster, identify the ones using the Kafka protocol and parse the message metadata. This script shows a sample of the most recent Kafka events in the cluster.
```
Comment (oazizi000) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: nit: Pixie automatically traces all messages flowing through your cluster, identifies the..., and parses the message...

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+
+> Pixie parses the request body and gives us information on the name of the topic, what partition it is producing to and the total size of the message set. Inspecting the response shows us if there's any error code coming back from the Kafka broker, base offset, log append time, etc.
+
+This script is useful for inspecting a specific request or filtering all requests by a certain command (you'll need to [edit the PxL script](https://docs.px.dev/using-pixie/using-live-ui/#write-your-own-pxl-scripts-edit-an-existing-script) to do that).
+
+## Producer Consumer Latency
+
+Producer consumer latency is the time between when a producer pushes a message to a topic to when it is fetched by a consumer.
+
+It's important to monitor this latency because many incidents begin with consumer lag.
```
Comment (oazizi000) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: Maybe also add a a comment about how large consumer lags one one component can mean it is out of sync with the other components.

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+<svg title='' src='use-case-tutorials/kafka/kafka-consumer-rebalancing.png'/>
+:::
+
+> The **Consumer Groups** table shows us that we now have 3 members in the `shipping` consumer group.
+
+> Each consumer rebalancing event consists of 1 `JoinGroup` and 1 `SyncGroup` request per consumer. In the default rebalance protocol, there's a stop-the-world effect on the whole consumer group when a rebalancing event happens.
+
+> The **Consumer Rebalancing Events** table shows the traced `JoinGroup` and `SyncGroup` requests.
+
+> The **Consumer Rebalancing Delay** table at the top calculates the delay between when the `JoinGroup` request was sent and when the `SyncGroup` request was received. This measures the stop-the-world delay mentioned above.
```
Comment (oazizi000) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md:182]: Should we have a summary?

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 20]
```diff
@@ -93,7 +93,7 @@ Let's inspect the raw Kafka requests flowing through the cluster.
 
 1. Select `px/kafka_data` from the script drop-down menu.
 
-> Pixie is able to [automatically trace](https://docs.px.dev/about-pixie/pixie-ebpf/) all messages flowing through your cluster, identify the ones using the Kafka protocol and parse the message metadata. This script shows a sample of the most recent Kafka events in the cluster.
+> Pixie [automatically traces](https://docs.px.dev/about-pixie/pixie-ebpf/) all messages flowing through your cluster, identify the ones using the Kafka protocol, and parses the message metadata. This script shows a sample of the most recent Kafka events in the cluster.
```
Comment (chengruizhe) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: nit: identify -> identifies

File: content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md [hunk 21]
```diff
@@ -113,9 +113,9 @@ This script is useful for inspecting a specific request or filtering all request
 
 ## Producer Consumer Latency
 
-Producer consumer latency is the time between when a producer pushes a message to a topic to when it is fetched by a consumer.
+Producer consumer latency is the time between when a producer pushes a message to a topic to when it is fetched by a consumer. This latency is important to monitor because many incidents begin with consumer lag.
 
-It's important to monitor this latency because many incidents begin with consumer lag.
+A large consumer lag in one component can mean that it is out of sync with other components. In our demo application, the `invoicing` service is not consuming messages from the `order` topic as quickly as the `shipment` service is. If you check the application front-end, you will see orders on the Shipment page that are not present on the Invoicing page. If not addressed, this application may fail to create invoices for all of the orders it receives and lose out on money owed for the orders that  successfully shipped.
```
Comment (chengruizhe) [inline content/en/04-tutorials/01-pixie-101/07-kafka-monitoring.md]: `shipment` service -> `shipping` service
There is an extra space between `that` and `successfully` on the last sentence.

**Review by chengruizhe — DISMISSED** [review]
Left comments on some typos.

**Review by oazizi000 — DISMISSED** [review]

**Review by aimichelle — APPROVED** [review]
